File: Project Enhancements/Brea Software Design and Engineering Enhancement/Brea_SoftDesignEngineeringEnhancement/Source/test/ContactServiceTest.cpp
#include "pch.h"
#include "CppUnitTest.h"
#include "Source/src/ContactService.h"

#include <string>

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

TEST_CLASS(ContactServiceTest) {
public:
    // Testing variables
    std::string testID = "12345";
    std::string testFirstName = "John";
    std::string testLastName = "Doe";
    std::string testPhone = "1234567890";
    std::string testAddress = "123 Main St";

    // Service instance
    ContactService service;

    // Test Add Contact Method
    TEST_METHOD(testAddContact) {
        service.addContact(testID, testFirstName, testLastName, testPhone, testAddress);
        
        Contact contact = service.getContact(testID);
        Assert::AreEqual(contact.getID(), testID);
        Assert::AreEqual(contact.getFirstName(), testFirstName);
        Assert::AreEqual(contact.getLastName(), testLastName);
        Assert::AreEqual(contact.getPhone(), testPhone);
        Assert::AreEqual(contact.getAddress(), testAddress);
    }

    // Test Remove Contact Method
    TEST_METHOD(testRemoveContact) {
        service.addContact(testID, testFirstName, testLastName, testPhone, testAddress);
        
        // Verify contact exists
        try {
            Contact contact = service.getContact(testID);
            Assert::IsTrue(true);
        }
        catch (...) {
            Assert::Fail(L"Unexpected exception");
        }

        // Remove contact
        service.removeContact(testID);

        // Verify contact was removed
        try {
            service.getContact(testID);
            Assert::Fail(L"Should have thrown exception for removed contact");
        }
        catch (const std::out_of_range&) {
            Assert::IsTrue(true);
        }
        catch (...) {
            Assert::Fail(L"Unexpected exception");
        }
    }

    // Test Update Methods
    TEST_METHOD(testUpdateMethods) {
        std::string newFirstName = "Jane";
        std::string newLastName = "Smith";
        std::string newPhone = "0987654321";
        std::string newAddress = "456 Oak Ave";

        service.addContact(testID, testFirstName, testLastName, testPhone, testAddress);
        
        service.updateContactFirstName(testID, newFirstName);
        service.updateContactLastName(testID, newLastName);
        service.updateContactPhone(testID, newPhone);
        service.updateContactAddress(testID, newAddress);

        Contact contact = service.getContact(testID);
        Assert::AreEqual(contact.getFirstName(), newFirstName);
        Assert::AreEqual(contact.getLastName(), newLastName);
        Assert::AreEqual(contact.getPhone(), newPhone);
        Assert::AreEqual(contact.getAddress(), newAddress);
    }

    // Test Add Contact Throws for ID
    TEST_METHOD(testAddContactThrowID) {
        auto emptyID = [this] { service.addContact("", "John", "Doe", "1234567890", "123 Main St"); };
        Assert::ExpectException<std::invalid_argument>(emptyID);

        auto longID = [this] { service.addContact("12345678901", "John", "Doe", "1234567890", "123 Main St"); };
        Assert::ExpectException<std::invalid_argument>(longID);
    }
};

File: Project Enhancements/Brea Database Enhancement/BreaCS320Proj1/src/test/ContactServiceTest.java
package test;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import main.ContactService;

class ContactServiceTest {

	@Test
	void testContactServiceAddContact() {
		ContactService contacts = new ContactService();
		contacts.addContact("12345", "Zach", "Brea", "1234567890", "123 Main St");
		assertAll("Should add Contact to contacts with no errors",
				() -> assertTrue(contacts.getContact("12345").getiD().equals("12345")),
				() -> assertTrue(contacts.getContact("12345").getFirstName().equals("Zach")),
				() -> assertTrue(contacts.getContact("12345").getLastName().equals("Brea")),
				() -> assertTrue(contacts.getContact("12345").getPhone().equals("1234567890")),
				() -> assertTrue(contacts.getContact("12345").getAddress().equals("123 Main St")));
	}
	
	@Test
	void testContactServiceIDNull() {
		ContactService contacts = new ContactService();
		Assertions.assertThrows(IllegalArgumentException.class, () -> {
			contacts.addContact("", "Zach", "Brea", "1234567890", "123 Main St");
		});
	}
		
	@Test
	void testContactServiceIDTooLong() {
		ContactService contacts = new ContactService();
		Assertions.assertThrows(IllegalArgumentException.class, () -> {
			contacts.addContact("12345678901", "Zach", "Brea", "1234567890", "123 Main St");
		});
	}
	
	@Test
	void testContactServiceUpdateContactFirstName() {
		ContactService contacts = new ContactService();
		contacts.addContact("12345", "Zach", "Brea", "1234567890", "123 Main St");
		
		contacts.updateContactFirstName("12345", "Zachary");
		
		assertTrue(contacts.getContact("12345").getFirstName().equals("Zachary"));
	}
	
	@Test
	void testContactServiceUpdateContactLastName() {
		ContactService contacts = new ContactService();
		contacts.addContact("12345", "Zach", "Brea", "1234567890", "123 Main St");
		
		contacts.updateContactLastName("12345", "Bread");
		
		assertTrue(contacts.getContact("12345").getLastName().equals("Bread"));
	}
	
	@Test
	void testContactServiceUpdateContactPhone() {
		ContactService contacts = new ContactService();
		contacts.addContact("12345", "Zach", "Brea", "1234567890", "123 Main St");
		
		contacts.updateContactPhone("12345", "0987654321");
		
		assertTrue(contacts.getContact("12345").getPhone().equals("0987654321"));
	}
	
	@Test
	void testContactServiceUpdateContactAddress() {
		ContactService contacts = new ContactService();
		contacts.addContact("12345", "Zach", "Brea", "1234567890", "123 Main St");
		
		contacts.updateContactAddress("12345", "321 1st Ave");
		
		assertTrue(contacts.getContact("12345").getAddress().equals("321 1st Ave"));
	}
	
	@Test
	void testContactServiceDeleteContact() {
		ContactService contacts = new ContactService();
		contacts.addContact("12345", "Zach", "Brea", "1234567890", "123 Main St");
		
		contacts.removeContact("12345");
		
		assertThrows(NullPointerException.class, () -> {
			contacts.getContact("12345").getiD().equals("12345");
		});
	}

}
